edgelet/edgelet-http-workload/src/edge_ca.rs

```diff
@@ -0,0 +1,411 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+#[cfg(not(test))]
+use aziot_cert_client_async::Client as CertClient;
+#[cfg(not(test))]
+use aziot_key_client_async::Client as KeyClient;
+#[cfg(not(test))]
+use aziot_key_openssl_engine as KeyEngine;
+
+#[cfg(test)]
+use test_common::client::CertClient;
+#[cfg(test)]
+use test_common::client::KeyClient;
+#[cfg(test)]
+use test_common::client::KeyEngine;
+
+pub(crate) struct EdgeCaRenewal {
+    rotate_key: bool,
+    temp_cert: String,
+    cert_client: std::sync::Arc<futures_util::lock::Mutex<CertClient>>,
+    key_client: std::sync::Arc<futures_util::lock::Mutex<KeyClient>>,
+    key_connector: http_common::Connector,
+    renewal_tx: tokio::sync::mpsc::UnboundedSender<edgelet_core::WatchdogAction>,
+}
+
+impl EdgeCaRenewal {
+    pub fn new(
+        rotate_key: bool,
+        config: &crate::WorkloadConfig,
+        cert_client: std::sync::Arc<futures_util::lock::Mutex<CertClient>>,
+        key_client: std::sync::Arc<futures_util::lock::Mutex<KeyClient>>,
+        key_connector: http_common::Connector,
+        renewal_tx: tokio::sync::mpsc::UnboundedSender<edgelet_core::WatchdogAction>,
+    ) -> Self {
+        let temp_cert = format!("{}-temp", config.edge_ca_cert);
+
+        EdgeCaRenewal {
+            rotate_key,
+            temp_cert,
+            cert_client,
+            key_client,
+            key_connector,
+            renewal_tx,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl cert_renewal::CertInterface for EdgeCaRenewal {
+    type NewKey = String;
+
+    async fn get_cert(
+        &mut self,
+        cert_id: &str,
+    ) -> Result<Vec<openssl::x509::X509>, cert_renewal::Error> {
+        let cert_client = self.cert_client.lock().await;
+
+        let cert = cert_client
+            .get_cert(cert_id)
+            .await
+            .map_err(|_| cert_renewal::Error::retryable_error("failed to retrieve edge CA cert"))?;
+
+        let cert_chain = openssl::x509::X509::stack_from_pem(&cert)
+            .map_err(|_| cert_renewal::Error::fatal_error("failed to parse edge CA cert"))?;
```

Comment:
Maybe one of these days we can make the certd client methods return openssl::X509 directly for convenience, and the keyd client methods return `(openssl::PKey<Private>, PKey<Public>)` directly. Well, the keyd client is a bit complicated because it would only be on the sync client, and it would need to be a separate sync client than the existing one because the existing one is itself used by the openssl engine. But the certd client at least would probably save a bunch of boilerplate everywhere.

Reply:
Yeah, definitely.